### DISCUSSION TITLE
Feature Request: Support for reading from and writing to encrypted Parquet files

### DISCUSSION THREAD

**Discussion by ravisharda** [duckdb/duckdb#6923] [Ideas]
Why do you want this feature?

We currently use DuckDB to merge records in incoming Parquet files into existing Parquet files, in one of our applications. Both sets of Parquet files are downloaded, then merged using DuckDB, and the results are outputted to new Parquet files. We use DuckDB's `read_parquet()` function for loading Parquet files and `COPY... TO` statement for sending the processed output to new Parquet files. 

Currently, those constructs don't support reading from/writing to encrypted files, which prevents us from performing application-level data-at-rest encryption. Data-at-rest encryption is a table-stakes requirement for us. 

Note:
1. The scope of data-at-rest encryption includes database file encryption, but that's already covered in https://github.com/duckdb/duckdb/discussions/4512. 
2. Has anyone tried loading and outputting encrypted Parquet files via DuckDB's [Apache Arrow import/export](https://duckdb.org/docs/api/java#arrow-methods) support? If that works, using Arrow streams may be a reasonable workaround.

**Comment by Alex-Monahan** [comment]
A coworker of mine is testing the pyarrow encryption/decryption method. I'll check with him if we can create a how to guide based on that approach once he gets it working!

That said, native DuckDB capability would be useful!

**Reply by ravisharda** [reply]
Arrow python docs have some examples of [reading (and writing)](https://arrow.apache.org/docs/python/parquet.html#parquet-modular-encryption-columnar-encryption) encrypted parquet files, but they don't exist for Java. 

I've asked a question in Apache Arrow repo: https://github.com/apache/arrow/issues/34851.

**Comment by hannes** [comment]
@ravisharda please ping me at hannes@duckblabs.com to discuss additional options

**Comment by ravisharda** [comment]
For folks watching this request, DuckDB v0.10.0 [introduced](https://duckdb.org/2024/02/13/announcing-duckdb-0100.html#new-features) support for Parqet file encryption.
